Recent changes for issue 1

ShahriyarR opened the pull request:
Fixes #1

Commit:
Changes for issue 1

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -108,52 +120,51 @@ class Manufacturer(Enum):
-def generate_qr_code(
-    name: str, birth: str, *manufacturer: Manufacturer, **date: str
+def generate(
+    name: str, birth: str, *manufacturers: Manufacturer, **dates: str
```
I still think it would be better to define it as
```
def generate(name: str, birth_date: str, **vaccinations: Manufacturer):
```
using dates a names in **kwargs. Eventually, we could support dict[date, Manufacturer] instead of plain **kwargs. See also my comment below.
I would approve a slight delay for adding support for these features since it would look really cool.

ShahriyarR replied:
This is postponed as per our discussion.

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -164,7 +175,7 @@ def generate_qr_code(
 
 The big change is we are going to accept variable length manufacturer names and dates will be passed as keyword arguments.
-This is quite intuitive isn't it? Thinking about CLI call, it will look like: `./cli generate-qr-code Shako 1989-10-24 pfizer pfizer 1st_dosage=2021-01-01 2nd_dosage=2021-06-01`
+This is quite intuitive isn't it? Thinking about CLI call, it will look like: `./cli qr-code generate Shako 1989-10-24 pfizer pfizer 1st_dosage=2021-01-01 2nd_dosage=2021-06-01`
```
You do not need ./cli here. Your entry point could be called for example green-badge generate ...
At least this is how we call it in Israel, or you may call vacination generate-qr ... or something like that. ./cli is just redundant.

Commit:
recent changes based on review

ShahriyarR commented:
@asterkin PTAL

ShahriyarR commented:
@shruti1807 @slichtma Please read the article here:
https://github.com/BstLabs/medium-articles/blob/issue-1/how_to_convert_python_functions/article.md

shruti1807 commented:
Hey Shako,

I am going through it . Will get back to you by tomorrow.

Thanks
Shruti

On Tue, Feb 8, 2022 at 6:05 PM Shahriyar Rzayev ***@***.***>
wrote:

> @shruti1807 <https://github.com/shruti1807> @slichtma
> <https://github.com/slichtma> Please read the article here:
>
> https://github.com/BstLabs/medium-articles/blob/issue-1/how_to_convert_python_functions/article.md
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BstLabs/medium-articles/pull/2#issuecomment-1032561451>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGHRWYJG6HTNJGKB2VVRTUTU2EEYTANCNFSM5NRVZHJA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ShahriyarR commented:
@shruti1807 I will push recent changes and ping you :)

Commit:
fully revisited to reflect the Asher's idea

ShahriyarR commented:
@shruti1807 updated. thanks.

ShahriyarR commented:
@asterkin PTAL

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+
+In original article following Python libraries were shown for building a *vaccination QR code* for you : argparse, Click, Typer, Docopt, Fire.
+
+We want to extend this article by using [dynacli](https://pypi.org/project/dynacli/).
```
Should be DynaCLI to preserve brand consistency

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+## What are we building?
+
+The core concept behind our changes is to ease down the process of building CLI applications by focusing only on Python code itself.
+To keep it simple dynacli generates help messages from the Python function docstrings and the function arguments are converted to CLI commands.
```
DynaCLI

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+
+Building CLIs in general divided in 2 steps: first the core code and second the CLI itself predefined as set of arguments, as shown above.
+
+With [dynacli](https://pypi.org/project/dynacli/) we can skip 2nd part by designing our functions to be CLI-friendly.
```
DynaCLI

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+
+## Our changes to the original code
+
+First off all, we would like to restructure the CLI - thinking about CLI there should be `./qr-code` then `green-badge` feature-set(actual Python package) it is for storing all commands,
```
better "grouping all command from one feature together" than "storing"

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+
+First off all, we would like to restructure the CLI - thinking about CLI there should be `./qr-code` then `green-badge` feature-set(actual Python package) it is for storing all commands,
+then `generate` to generate actual QR codes:
+
```
You discussed a feature of using dynacli init to bootstrap a new CLI project. I assume it's not there yet.

ShahriyarR replied:
Nope, it is not even started.

asterkin commented on how_to_convert_python_functions/article.md:
````diff
@@ -0,0 +1,366 @@
+
+```py
+class Vaccination(TypedDict):
+    date: datetime
````
date class would be enough; we do not store time

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+
+class QRCode(TypedDict):
+    name: str
+    birth: str
```
better birth_date

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+class QRCode(TypedDict):
+    name: str
+    birth: str
+    vaccination: list[Vaccination]
```
should be vaccinations (plural)

asterkin commented on how_to_convert_python_functions/article.md:
````diff
@@ -0,0 +1,366 @@
+The second big step is to design actual `generate` function in a way that it is going to accept all necessary information as arguments:
+
+```py
+def generate(first_name: str, last_name: str, birthdate: str,  **vaccination: str) -> None:
````
vaccinations (plural)

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+    """
+    qr_code = QRCode(
+        name=f"{first_name} {last_name}",
+        birth=birthdate,
```
birth_date, vaccinations

asterkin commented on how_to_convert_python_functions/article.md:
````diff
@@ -0,0 +1,366 @@
+```
+
+Or we can use [Literal](https://docs.python.org/3.9/library/typing.html#typing.Literal) -  and I am going to use it here in favor of Enum, by that we lose some guard but the clean interface is always beats and you can always add extra checks outside of the interface.
+
````
In this case, supporting Enums would be nicer and I did not get what you benefit from literals. I do not object supporting them in DynaCLI, but we have to be clear and consistent.

ShahriyarR replied:
But in your edition the need for Enum is just eliminated - i.e we can not use Enum type with **kwargs in current DynaCLI implementation. That's why the need to the Enum type just disappeared in this article.

asterkin replied:
I do not understand why we cannot use Enum or any type in **kwargs: <type> and if this is the case, let's fix it. The whole idea was that the **kwargs could be of any type and you control it (yes, not argparse, but the user will not notice). What I'm missing?

ShahriyarR replied:
The problem is `**kwargs: EnumType` -> DynaCLI originally registers Enum type as choices and we did it either *args or a positional argument, and with `date=[choices]` it does not work.

asterkin commented on how_to_convert_python_functions/article.md:
```diff
@@ -0,0 +1,366 @@
+![QR](./code/qr_code.png)
+
+That's it. We have focused only to the core functionality, cleaned it up a bit, reshaped it to be CLI friendly - and that is how the CLIs should be.
+You only enjoy your Python code, nothing more.
```
As we discussed, you need to add a small paragrpah on Open Closed principle justifying keeping each command in a separate module.

ShahriyarR replied:
I guess it is already there

slichtma replied:
I can propose a number of polishing edits to the green pass/dynacli article. @ShahriyarR 
, is the doc good for me to pull, or do you need to merge outstanding edits first?

shruti1807 committed:
Update article.md

Small modifications in the language of the article.

shruti1807 committed:
Update article.md

shruti1807 committed:
Update article.md

slichtma committed:
Update article.md

Scott made edits on top of Shruti's . Ready for Shako's final review then posting on Medium under his name.  
Shako, when posting on medium pay special attention to the amount of text that fits in the title and first paragraph, when you look at the article in preview/surf-the-article-list mode. If it gets cut-off in the middle of an important phrase, consider rearranging the the text a bit. I can help if needed. All yours. When published, we'l target Hackernoon for re-publication.

slichtma committed:
Update article.md

slichtma committed:
Update article.md

Commit:
recent code changes

Commit:
Getting Shruti's and Scott's changes

ShahriyarR commented:
I have grabbed @shruti1807 and @slichtma changes and merged to my branch.
Now the text should be okay - remaining thing is to make code changes to our sample app.
The rest is good to go. Thanks for your help guys you are awesome :)

Commit:
Recent changes with updated code

Commit:
Code changes based on Asher's input

ShahriyarR commented:
@asterkin @shruti1807 @slichtma 
This is a updated version with small additions - I have merged your text as well.
Please review.

Commit:
Added recent code changes

slichtma committed:
Update article.md

One word, commmands, had 3 'm's in t.

ShahriyarR commented:
If we are good to go, please approve this PR and then I need to merge this PR, make this repo public.
Finally we need to publish this article and prepare next ones.
@asterkin @slichtma @shruti1807

slichtma committed:
Update article.md

The text looks good all around. To post on medium, we'll add a stock photo.

slichtma reviewed (approved):
Text all looks good/

asterkin commented on how_to_convert_python_functions/code/green_badge/generate.py:
```diff
@@ -0,0 +1,59 @@
+        name=f"{first_name} {last_name}",
+        birth=birth_date,
+        vaccinations=[
+            Vaccination(manufacturer=manufacturer, date=datetime.strptime(date, "%Y-%m-%d"))
```
Ideally, it should print a nice error message if one of vaccination pairs contains wrong or duplicate date. But you could leave it as a comment to keep the code simple.

Commit:
added todo message and clarification

Commit:
Merge branch 'issue-1' of github.com:BstLabs/medium-articles into issue-1
